### PR TITLE
Revert back to non-oc execution for read-only trxs

### LINKED ISSUE #1119: further improvement of parallel read-only trx execution fallback when OC tier up enabled
https://github.com/AntelopeIO/leap/issues/970 reported an issue that if OC compiled code was not available,
OC execution would fall back to thread-unsafe default EOS-VM-Interpreter or EOS-VM-JIT execution.
Due to time constraints, a workaround was implemented implemented to retry the execution in the future round. 
This is not optimal, as it might take a while before the compile is done and the user will wait for
non-deterministic time to get the transaction result.

The correct solution is to at `nodeos` startup, if EOS-VM-OC is enabled,
EOS-VM-OC, EOS-VM-Interpreter, and EOS-VM-JIT all need to be set up to support parallel execution.
In OC execution, if the OC compiled code is not available, the execution falls back to run parallelly in
EOS-VM-Interpreter or EOS-VM-JIT.

### PR DESCRIPTION
**Pull request by heifner**
When `eos-vm-oc-enable` is `auto` or `all` revert back to non-oc tier up when oc compile of read-only trx not ready.
* Refactors `eosvmoc_tier` into `wasm_interface_collection` so that all threads share the same `eosvmoc_tier` instead of having one per `wasm_interface`.
* Removes unneeded eos-vm fast shutdown.
* Fixes `substitute_apply` so it does not have to be set on each thread.

Resolves #1119

### CODE REVIEW

File: libraries/chain/wasm_interface_collection.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+#ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
+   if (eosvmoc) {
+      cb = [&](const digest_type& code_hash, uint8_t vm_version) {
+         eosvmoc->cc.free_code(code_hash, vm_version);
+      };
+   }
+#endif
+   wasmif.current_lib(lib, cb);
+   for (auto& w : threaded_wasmifs) {
+      w.second->current_lib(lib, cb);
```
Comment (spoonincode) [inline libraries/chain/wasm_interface_collection.cpp:95]: (I haven't studied this PR closely yet so apologies if this should be obvious..)

This part feels a little strange to me: there should only be a single OC code cache that is shared across all threads. So why pass the `cb` here that ` eosvmoc->cc.free_code(code_hash, vm_version)` per thread (presumably it gets called more than once).

Comment (heifner) [reply libraries/chain/wasm_interface_collection.cpp:95]: Each `wasm_interface` has its own `wasm_instantiation_cache`. There is only one `eosvmoc`. This makes sure that all the codes are freed no matter which `wasm_interface`s caches it.

**Review by greg7mdp — APPROVED** [review]

File: libraries/chain/include/eosio/chain/controller.hpp [hunk 2]
```diff
@@ -348,7 +349,7 @@ namespace eosio { namespace chain {
          */
 
          const apply_handler* find_apply_handler( account_name contract, scope_name scope, action_name act )const;
-         wasm_interface& get_wasm_interface();
+         wasm_interface_collection& get_wasm_interface();
```
[the remark this answers is not in the document]
Comment (heifner) [reply libraries/chain/include/eosio/chain/controller.hpp:352]: I consider `wasm_interface_collection` providing the composite pattern for `wasm_interface`s which why I left the name `get_wasm_interface`.

File: libraries/chain/wasm_interface.cpp [hunk 3]
```diff
@@ -74,66 +72,24 @@ namespace eosio { namespace chain {
...
-            once_is_enough = true;
-         }
-         if(cd) {
-            if (!context.is_applying_block()) // read_only_trx_test.py looks for this log statement
-               tlog("${a} speculatively executing ${h} with eos vm oc", ("a", context.get_receiver())("h", code_hash));
-            my->eosvmoc->exec->execute(*cd, my->eosvmoc->mem, context);
-            return;
-         }
-         else if (context.trx_context.is_read_only()) {
-            if (failure == chain::eosvmoc::code_cache_base::get_cd_failure::temporary) {
```
[the remark this answers is not in the document]
Comment (heifner) [reply libraries/chain/wasm_interface.cpp:115]: Do you think they should be removed? Seems like they might provide some usefulness in the future.

[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/wasm_interface.cpp:115]: Those failure reasons were introduced specifically for the temporary workaround. OK, let us keep them as they provide exact failed reasons.

File: libraries/chain/wasm_interface_collection.cpp [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+   EOS_ASSERT(!is_on_main_thread(), misc_exception, "init_thread_local_data called on the main thread");
+#ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
+   if (is_eos_vm_oc_enabled()) {
+      // EOSVMOC needs further initialization of its thread local data
+      if (eosvmoc)
+         eosvmoc->init_thread_local_data();
+      wasmif.init_thread_local_data();
+   } else
+#endif
+   {
```
[the remark this answers is not in the document]
Comment (heifner) [reply libraries/chain/wasm_interface_collection.cpp:117]: I thought we had integration tests that would cover it. Turns out they were executing all trxs on the main thread. I'm trying to tweak the test to verify the fall back case.

I hard-coded a eosvmoc failure and this does indeed crash. Great catch!.

[the remark this answers is not in the document]
Comment (heifner) [reply libraries/chain/wasm_interface_collection.cpp]: Test added that covers this.

**Review by linh2931 — APPROVED** [review]

File: libraries/chain/wasm_interface_collection.cpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+            tlog("${a} speculatively executing ${h} with eos vm oc", ("a", context.get_receiver())("h", code_hash));
+         eosvmoc->exec->execute(*cd, eosvmoc->mem, context);
+         return;
+      }
+   }
+#endif
+   if (is_on_main_thread()) {
+      wasmif.apply(code_hash, vm_type, vm_version, context);
+      return;
+   }
```
Comment (linh2931) [inline libraries/chain/wasm_interface_collection.cpp:82]: why not an else here?

Comment (heifner) [reply libraries/chain/wasm_interface_collection.cpp:82]: The rest of the method short-cuts out with `return`. So this follows the rest of the method approach.

**Pull request merged**